diffrax/_term.py

```diff
@@ -776,3 +785,136 @@ def _to_vjp(_y, _diff_args, _diff_term):
+        if isinstance(control, AbstractBrownianIncrement):
+            control = control.W
+
+        dv = vf
+        dw = control
+        v_out = jtu.tree_map(operator.mul, dv, dw)
+        x_out = jtu.tree_map(jnp.zeros_like, v_out)
+        return x_out, v_out
+
+
+def _broadcast_pytree(source, target_tree_shape):
+    # Requires that source is a prefix tree of target_tree_shape
+    def _inner_broadcast(_src_arr, _inner_tree_shape):
+        _arr = jnp.asarray(_src_arr)
+        return jtu.tree_map(
+            lambda _leaf: jnp.broadcast_to(_arr, _leaf.shape), _inner_tree_shape
+        )
+
+    return jtu.tree_map(_inner_broadcast, source, target_tree_shape)
+
+
+class _LangevinDriftTerm(AbstractTerm):
+    gamma: _LangevinX
+    u: _LangevinX
+    grad_f: Callable[[_LangevinX], _LangevinX]
+
+    def vf(self, t: RealScalarLike, y: _LangevinTuple, args: Args) -> _LangevinTuple:
+        x, v = y
+        f_x = self.grad_f(x)
+        d_x = v
+        d_v = jtu.tree_map(
+            lambda _gamma, _u, _v, _f_x: -_gamma * _v - _u * _f_x,
+            self.gamma,
+            self.u,
+            v,
+            f_x,
+        )
+        d_y = (d_x, d_v)
+        return d_y
+
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> RealScalarLike:
+        return t1 - t0
+
+    def prod(self, vf: _LangevinTuple, control: RealScalarLike) -> _LangevinTuple:
+        return jtu.tree_map(lambda _vf: control * _vf, vf)
+
+
+def make_langevin_term(
+    gamma: PyTree[ArrayLike],
+    u: PyTree[ArrayLike],
+    grad_f: Callable,
+    bm: AbstractBrownianPath,
+    x0: _LangevinX,
+) -> MultiTerm[tuple[_LangevinDriftTerm, _LangevinDiffusionTerm]]:
+    r"""Creates a term that represents the Underdamped Langevin Diffusion, given by:
+    $d \mathbf{x}_t = \mathbf{v}_t dt$
+    $d \mathbf{v}_t = - \gamma \mathbf{v}_t dt - u
+    \nabla f( \mathbf{x}_t ) dt + \sqrt{2 \gamma u} d W_t.$
+    where $\mathbf{x}_t, \mathbf{v}_t \in \mathbb{R}^d$ represent the position
+    and velocity, $W$ is a Brownian motion in $\mathbb{R}^d$,
+    $f: \mathbb{R}^d \rightarrow \mathbb{R}$ is a potential function, and
+    $ \gamma,u\in\mathbb{R}^{d\times d}$ are diagonal matrices governing
+    the friction and the dampening of the system.
+
+    **Arguments:**
+
+    - `gamma`: A vector containing the diagonal entries of the friction matrix;
+        a PyTree of the same shape as `x0`.
+    - `u`: A vector containing the diagonal entries of the dampening matrix;
+        a PyTree of the same shape as `x0`.
+    - `grad_f`: A callable representing the gradient of the potential function $f$.
+        This callable should take a PyTree of the same shape as `x0` and
+        return a PyTree of the same shape.
+    - `bm`: A Brownian path representing the Brownian motion $W$.
+    - `x0`: The initial state of the system (just the position without velocity);
+        only needed to check the PyTree structure of the other arguments.
```

Comment:
Is this argument really necessary?

Note that elsewhere in Diffrax we don't take `y0` as an input to our terms -- rather, we allow our term definitions to be polymorphic over many possible `y0`s. E.g. `ODETerm(lambda t, y, args: -y)` does not say anything about the shape of `y`.

I suspect the same should probably be doable here by moving the broadcasting into the new terms themselves.

Reply:
I rely on gamma to be the right shape for the coefficients in `LangevinSRK`. So there are several options I see:
- recast `gamma` and `u` into the correct shapes in `LangevinSRK.init` and hold them in solver state (honestly might not be so bad)
- assume that `gamma` will have the same shape as `x0` (which I don't want to require, since it should be fine to have a scalar `gamma` even if `x` is a weird PyTree)
- do what is currently done.

To me the last option seems a bit more robust a bit more transparent to the user: if their arguments are of incorrect shapes, then a Langevin term cannot be made in the first place. Whereas if a solver complains about some PyTree mismatch, then it might not be as obvious what went wrong (or at least I'd have to fiddle with `try, catch` to provide clear error messages, etc). And I doubt this hinders the usability, since they'll need to have an `x0` somewhere anyway. But I am happy to go with the first option.

Reply:
I think the first option is probably my favourite! I'd like to keep this abstraction the way it is.

(The reason it is the way it is, by the way, is because of cases like https://docs.kidger.site/diffrax/examples/neural_ode/ -- we often want to be able to describe the vector field at `__init__` time whilst only obtaining the initial condition at `__call__` time.)

Reply:
Sounds good!